### PR TITLE
Further optimization of the SiPixel LA PCL workflow

### PR DESCRIPTION
**Pull request by mmusich**
#### PR description:

This PR builds on top of https://github.com/cms-sw/cmssw/pull/36538 and proposes some further optimization and bug-fixes to the   SiPixel Lorentz Angle PCL workflow:
   * further memory footprint reduction is achieved in commit 69b9256f3d0be42121c826f63094094fee22375a by loading the `SiPixelTemplate` object in memory on demand in the case in which the module `SiPixelLorentzAnglePCLWorker` is not run in PCL mode (`notInPCL_ = True`) and by reducing the number of bins of the 2D drift vs depth maps.
   * for Run2 replays the list of "new" modules is different. That is dealt with via era modifiers in commit c9aa27ab2e045b958c5294d753eaee9ccae6a8a0
   * a bug that was messing up the computation of the value of the Lorentz Angle for the "new" modules is addressed at fb4f1262f51a74929275b33e29e4d5c795b84bda
   * monitoring of the input number of clusters and output Lorentz Angle (and fit chi2/ndf) is introduced at 0cfb6bf446943c82b59b37bd66c4062fd3b8c080
   * `std::endl` at the end of LogMessages is removed as requested at https://github.com/cms-sw/cmssw/pull/36538#issuecomment-998643023 in commit 0cfb6bf446943c82b59b37bd66c4062fd3b8c080
   * some documentation is added in commit 199bcd689cdd6619718a85d496022abb7e5031e8
    
#### PR validation:

Privately run:

```console
 cmsDriver.py testReAlCa -s ALCA:PromptCalibProdSiPixelLorentzAngle --conditions 121X_dataRun3_Express_TIER0_REPLAY_Run2_v1 --scenario pp --data --era Run2_2018 --datatier ALCARECO --eventcontent ALCARECO --processName=ReAlCa -n 100000 --dasquery='file dataset=/StreamExpress/Tier0_REPLAY_2021-SiPixelCalSingleMuon-Express-v1/ALCARECO' --customise_commands='process.ALCARECOCalSignleMuonFilterForSiPixelLorentzAngle.TriggerResultsTag = cms.InputTag ( "TriggerResults","","HLT" ) ; process.ALCARECOCalSignleMuonFilterForSiPixelLorentzAngle.HLTPaths = ["*"]' --nThreads=4 
```

followed by:

```console
 cmsDriver.py stepHarvest -s ALCAHARVEST:SiPixelLA --conditions 121X_dataRun3_Express_TIER0_REPLAY_Run2_v1 --scenario pp --data --era Run2_2018 --filein file:PromptCalibProdSiPixelLorentzAngle.root -n -1
```
the results are uploaded to a private DQM GUI at https://tinyurl.com/yxcmsvn6:
   * necessitates tunneling (`ssh -NL 8060:localhost:8060 <USER>@lxplus723.cern.ch`). 

Profiling the RSS vs time of the job containing the `ALCA:PromptCalibProdSiPixelLorentzAngle` (run on 4 threads and 4 stream) shows a reduction of about half:

memory_target

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Not a backport, but might be backported together with https://github.com/cms-sw/cmssw/pull/36538 to CMSSW_12_2_X as agreed at https://github.com/cms-sw/cmssw/pull/36538#issuecomment-997386926

### CODE REVIEW

**Comment by tvami** [comment]
@cmsbuild , plese test

**Comment by mmusich** [comment]
type bugfix

**Comment by mmusich** [comment]
please test

there is a typo at https://github.com/cms-sw/cmssw/pull/36565#issuecomment-998982330

**Comment by yuanchao** [comment]
+1
* memory optimization and bug fixes
* matrix tests and unit tests passed

File: CalibTracker/SiPixelLorentzAngle/src/SiPixelLorentzAnglePCLWorker.cc [hunk 1]
```diff
@@ -845,7 +881,7 @@ void SiPixelLorentzAnglePCLWorker::fillDescriptions(edm::ConfigurationDescriptio
   desc.add<double>("clustChargeMaxPerLength", 50000)
       ->setComment("maximum cluster charge per unit length of pixel depth (z)");
   desc.add<int>("binsDepth", 50)->setComment("# bins for electron production depth axis");
-  desc.add<int>("binsDrift", 200)->setComment("# bins for electron drift axis");
+  desc.add<int>("binsDrift", 100)->setComment("# bins for electron drift axis");
```
Comment (perrotta) [inline CalibTracker/SiPixelLorentzAngle/src/SiPixelLorentzAnglePCLWorker.cc:884]: This is not modified in CalibTracker/SiPixelLorentzAngle/python/SiPixelLorentzAnglePCLWorker_cfi.py, and therefore this change won't have ay practical effect on the actual configs. Is it intended?

File: CalibTracker/SiPixelLorentzAngle/python/SiPixelLorentzAnglePCLWorker_cfi.py [hunk 2]
```diff
@@ -6,14 +6,7 @@
...
-                                "BPix_BmO_SEC3_LYR2_LDR5F_MOD1",
-                                "BPix_BmO_SEC3_LYR2_LDR5F_MOD2",
-                                "BPix_BmO_SEC3_LYR2_LDR5F_MOD3",
-                                "BPix_BpO_SEC1_LYR2_LDR1F_MOD1",
-                                "BPix_BpO_SEC1_LYR2_LDR1F_MOD2",
-                                "BPix_BpO_SEC1_LYR2_LDR1F_MOD3"),
+    newmodulelist = cms.vstring(),
     src = cms.InputTag("TrackRefitter"),
     binsDepth    = cms.int32(50),
     binsDrift =    cms.int32(200),
```
Comment (perrotta) [inline CalibTracker/SiPixelLorentzAngle/python/SiPixelLorentzAnglePCLWorker_cfi.py:12]: Perhaps you can clone SiiPixelLorentzAnglePCLWorker from the siPixelLorentzAnglePCLWorker generated by the fillDescription method, modifying only the parameters which have to be changer with respect to the default ones (here `binsDrift`, unless there is a typo somewhere)

**Comment by perrotta** [comment]
@mmusich, even if not part of this PR, the static analyzer reports about a non const static variable `lower_bin_` in CalibTracker/SiPixelLorentzAngle/src/SiPixelLorentzAngle.cc
As far as I can see, that variable is never used in that code: am I missing something, or its definition can simply get removed from SiPixelLorentzAngle.cc?
(Not necessarily for this PR, unless given the comments above you realize that an update is needed anyhow)

File: CalibTracker/SiPixelLorentzAngle/src/SiPixelLorentzAnglePCLWorker.cc (hunk 1 above)
Comment (mmusich) [reply CalibTracker/SiPixelLorentzAngle/src/SiPixelLorentzAnglePCLWorker.cc:877]: done at https://github.com/cms-sw/cmssw/pull/36565/commits/3b421bc3a014a8c63325bc4c6448220c2d073621

File: CalibTracker/SiPixelLorentzAngle/python/SiPixelLorentzAnglePCLWorker_cfi.py (hunk 2 above)
Comment (mmusich) [reply CalibTracker/SiPixelLorentzAngle/python/SiPixelLorentzAnglePCLWorker_cfi.py]: done at https://github.com/cms-sw/cmssw/pull/36565/commits/3b421bc3a014a8c63325bc4c6448220c2d073621

**Comment by mmusich** [comment]
> As far as I can see, that variable is never used in that code: am I missing something, or its definition can simply get removed from SiPixelLorentzAngle.cc?

done at https://github.com/mmusich/cmssw/commit/95f81f2e899a6fbf82d7c687cc2a43b4c039ec54

File: CalibTracker/SiPixelLorentzAngle/src/SiPixelLorentzAnglePCLHarvester.cc [hunk 3]
```diff
@@ -231,16 +258,37 @@ void SiPixelLorentzAnglePCLHarvester::dqmEndJob(DQMStore::IBooker& iBooker, DQMS
...
+                     "tan#theta_{LA}/B by sector ;pixel sector;measured tan(#theta_{LA})/B [1/T]",
+                     maxSect,
+                     -0.5,
+                     maxSect + 0.5);
+  MonitorElement* h_bySectChi2_ = iBooker.book1D(
+      "h_bySectorChi2", "#chi^{2}/ndf by sector;pixel sector; fit #chi^{2}/ndf", maxSect, -0.5, maxSect + 0.5);
+
+  // copy the bin labels from the occupancy histogram
+  for (int bin = 1; bin < maxSect; bin++) {
+    const auto& binName = hists.h_bySectOccupancy_->getTH1()->GetXaxis()->GetBinLabel(bin);
```
Comment (mmusich) [inline CalibTracker/SiPixelLorentzAngle/src/SiPixelLorentzAnglePCLHarvester.cc:287]: is this tolerated / thread safe? @cms-sw/dqm-l2 
I didn't find a `MonitorElement::getBinLabel()` method at https://github.com/cms-sw/cmssw/blob/master/DQMServices/Core/interface/MonitorElement.h

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
please test

**Comment by mmusich** [comment]
@cms-sw/alca-l2 ping

**Comment by tvami** [comment]
Hi @mmusich , Happy New Year! I thought you had a question to DQM in this PR. Is that resolved? Thanks

**Comment by mmusich** [comment]
> Is that resolved?

As far as I am concerned, yes

**Comment by tvami** [comment]
Ok, from the AlCa side I think we are good

**Comment by tvami** [comment]
+alca

**Pull request merged**